### ISSUE TITLE
Trainer

### ISSUE THREAD

**Issue by xrsrke** [xrsrke/pipegoose#18]
"Implement a Trainer, which is a wrapper of low-level `DataParallel`, `TensorParallel` and `PipelineParallel` modules. The user just plugs in their model and dataloader and trains. Similar to `transformers`.

```
trainer = Trainer(
	model=model,
	train_dataloader=train_dataloader,
  eval_dataloader=eval_dataloader,
  optimizers=optim,
  num_epochs=10,
	callbacks=[LearningRateFinder()]
)

trainer.fit()
```

**Comment by isamu-isozaki** [comment]
I think I'll do this tonight since it seems the easiest

**Comment by xrsrke** [comment]
@isamu-isozaki Awesome, thank you! I will get back to you in a few hours with all the details!!

**Comment by isamu-isozaki** [comment]
@xrsrke I was thinking of maybe just inheriting from transformer's Trainer. wdyt?

**Comment by xrsrke** [comment]
@xrsrke  No, I just checked the Trainer from Transformers. They modified our model's devices and stuff. We prefer implementing our own so we can incorporate distributed logging and callback in a specific rank, ParallelMode... and future changes. I just added some demo code ([link](https://github.com/xrsrke/pipegoose/commit/46e1b0cf8bf00c4a3ce1642faa85549593fc727e)).

Also one note, we only apply a specific parallel mode based on the parallel_context. For example, if data_parallel_size is greater than 1, then we wrap the model with DataParallel.
